cstore_writer.c

```diff
@@ -310,104 +323,228 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
+	{
+		startingBlock = 1;
+	}
+	else
+	{
+		uint32 footerHeaderDataLength =
+			footerHeaderBufferPageHeader->pd_lower- SizeOfPageHeaderData;
+		DeserializeTableFooterMetadata(footerHeaderPageContent,
+									  footerHeaderDataLength,
+									  &existingFoooterStartingBlock,
+									  &existingFooterBlockCount);
+
+		/*
+		 * We here we decide where to start writing new table footer metadata.
+		 * If deserialization fail for any reason starting block is set to an
+		 * invalid starting block number (0).
+		 *
+		 * If there is a parsing error we start from the first buffer (1). If
+		 * it is fine we check if there are enough empty buffers to accommodate
+		 * all blocks of table footer metadata and set new starting buffer to
+		 * first buffer (1). If there are not enough buffers we start writing
+		 * after the all used buffers.
+		 */
+		if (existingFoooterStartingBlock == 0)
+		{
+			startingBlock = 1;
+		}
+		else if (footerDataBlockCount < existingFoooterStartingBlock)
+		{
+			startingBlock = 1;
+		}
+		else
+		{
+			startingBlock = existingFoooterStartingBlock + existingFooterBlockCount;
+		}
+	}
+
+	dataOffset = 0;
+
+	for (blockIndex = 0; blockIndex < footerDataBlockCount; blockIndex++)
+	{
+		Buffer buffer = InvalidBuffer;
+		BlockNumber blockNumber = blockIndex + startingBlock;
+		Page page;
+		PageHeader pageHeader = NULL;
+		char * pageData = NULL;
+		int copySize = blockDataSize;
+
+		if (blockNumber >= forkBlockCount)
+		{
+			blockNumber = P_NEW;
+		}
 
-	SyncAndCloseFile(tableFooterFile);
+		buffer = ReadBufferExtended(relation, FOOTER_FORKNUM, blockNumber,
+								    RBM_NORMAL, NULL);
 
-	pfree(tableFooterBuffer->data);
-	pfree(tableFooterBuffer);
-	pfree(postscriptBuffer->data);
-	pfree(postscriptBuffer);
+		LockBuffer(buffer, BUFFER_LOCK_EXCLUSIVE);
+
+		START_CRIT_SECTION();
+		page = BufferGetPage(buffer);
+
+		PageInit(page, BLCKSZ, 0);
+		pageData = PageGetContents(page);
+
+		if ( (dataLength - dataOffset) <= blockDataSize)
+		{
+			copySize = dataLength - dataOffset;
+		}
+
+		pageHeader = (PageHeader) page;
+		pageHeader->pd_lower = SizeOfPageHeaderData + copySize;
+		memcpy(pageData, wholeFooter->data + dataOffset, copySize);
+
+		MarkBufferDirty(buffer);
+
+		if (loggingEnabled)
+		{
+			LOG_NEWPAGE_BUFFER(buffer);
+		}
+
+		END_CRIT_SECTION();
+
+		UnlockReleaseBuffer(buffer);
+
+		dataOffset += copySize;
+
+	}
+
+	/* all table footer is written, update the header page */
+	tableFooterMetadata = SerializeTableFooterMetadata(startingBlock, footerDataBlockCount);
+
+	START_CRIT_SECTION();
+	PageInit(footerHeaderPage, BLCKSZ, 0);
+	memcpy(footerHeaderPageContent, tableFooterMetadata->data, tableFooterMetadata->len);
+	footerHeaderBufferPageHeader->pd_lower = SizeOfPageHeaderData + tableFooterMetadata->len;
+	MarkBufferDirty(footerHeaderBuffer);
+
+	/*
+	 * Changes to headerBuffer is logged regardless of logging setting to enable
+	 * recovery after crash.
+	 */
+	LOG_NEWPAGE_BUFFER(footerHeaderBuffer);
```

Comment:
I originally thought it would be nicer for crash recovery to log changes to metadata regardless of logging setting